### PR TITLE
Team/Teammate Matching

### PR DESCRIPTION
**Pull request by chloetanlimco**
image

What's done: 
- Simple teammate recommendation route - filters hackers by year
- Frontend (for the most part) for teammate recommendations

Need to do: 
- Add a migration for Team model to have a property status to indicate whether the team is looking for members
- Show team or teammate recommendations based on team membership & whether the team is looking for members
- Onclick functions / add routes to invite user / request to join team -> maybe this can just pull up an email / slack message? then we don't need to add additional properties to the models 
- Build out frontend for team recommendations

### CODE REVIEW

File: api/people.js [hunk 1]
```diff
@@ -56,4 +56,18 @@ router.put("/houses", async (req, res) => {
   }
 });
 
+router.delete("/houses/:id", async (req, res) => {
```
Comment (MaxLeiter) [inline api/people.js:59]: Seems like a commit of Andreas's snuck in

File: api/teamMatching.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
+const express = require("express");
```
Comment (MaxLeiter) [inline api/teamMatching.js:1]: if we can use `import` statements here let's use them

File: api/teamMatching.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  });
+
+  const hackerProfiles = await models.HackerProfile.findAll({
+    where: {
+      year: curruser.year,
+      teamId: null,
+      status: "confirmed",
+      role: "hacker",
+    },
+    // include: [
```
Comment (MaxLeiter) [inline api/teamMatching.js:25]: Can we re-add `attributes` so we aren't sending all the fields?

File: lib/live.ts [hunk 4]
```diff
@@ -47,8 +47,11 @@ async function updateTask(updatedTask) {
   return result.status === 200;
 }
 
-async function deleteTask(task) {
-  const urlRoute = "/api/points/tasks/" + task.id;
+async function deleteTask(req, task) {
```
Comment (MaxLeiter) [inline lib/live.ts:50]: also from andreas

File: pages/api/teamMatching.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+// GET /api/matching
+// - If a hacker is on a team, get that team info
+router.get("/", async (req, res) => {
+  const curruser = await models.HackerProfile.findOne({
+    where: {
+      userId: req.user.id,
+    },
+  });
+
+  const hackerProfiles = await models.HackerProfile.findAll({
```
Comment (MaxLeiter) [inline pages/api/teamMatching.ts:18]: same comments as api/teamMatching

**Review by MaxLeiter — APPROVED** [review]

**Review by MaxLeiter — CHANGES_REQUESTED** [review]
Didn't mean to approve before. The code seems good

**Comment by chloetanlimco** [comment]
image

**What's done:**

- Simple teammate recommendation route - filters hackers by year
- Frontend (for the most part) for teammate recommendations
- Frontend for team recommendations
- Added a migration for Team model to have details & status to indicate whether team is looking for members
- Set up many:many relationship between Team & HackerProfile to store team/teammate requests
- Onclick to request to join team




**Need to do:**

Team Management Page
- Edit blurb / details
- Show Teammate Suggestions
- Toggle whether they're looking for members
- Show people who've requested to join the team

ProjectTeam / Team Page (???)
- Allow people looking for teams to add url (LinkedIn / Portfolio) & display that on their teammate recommendation
- Show team or teammate recommendations based on team membership & whether the team is looking for members
- Remove requested teams from Suggested Teams List 
- Fix Join & Create functionality on /ProjectTeam (??)

MISC
- Remove Andreas' code from this branch

**Comment by chloetanlimco** [comment]
image
image
image

File: components/Team.tsx [hunk 6]
```diff
@@ -132,6 +169,48 @@ const Team = ({ team, profile }: Props) => {
...
+          <Header>Edit Team Visibility</Header>
+          <ChangeProfileOption>
+            <input
+              name="visibility"
+              type="checkbox"
+              checked={visible}
+              onChange={handleChangeVisibility}
+            />
+            <ShareProfileText>
+              Share the team name and description with hackers looking for teams!
```
Comment (MaxLeiter) [inline components/Team.tsx:183]: Can we change this to `Share the team name and description with hackers looking to join a team`?

Comment (MaxLeiter) [inline components/Team.tsx:183]: Can we change this to Share the team name and description with hackers looking to join a team?

File: components/Team.tsx [hunk 7]
```diff
@@ -132,6 +169,48 @@ const Team = ({ team, profile }: Props) => {
...
+            </ShareProfileText>
+          </ChangeProfileOption>
+        </Column>
+        <Column flexBasis={48}>
+          <Header>Edit Team Description</Header>
+          <Form>
+            <FormGroup>
+              <label>Displayed to potential hackers</label>
```
Comment (MaxLeiter) [inline components/Team.tsx:191]: I don't know if we need the label if we have the message next to the checkbox

File: components/TeammateSuggestions.tsx [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+                      <Subheading>
+                        {suggestion.year ? suggestion.year : "year unavailable"}
+                      </Subheading>
+                    </>
+                  )
+                }
+
+                
+                <Subheading>
+                  {suggestion.major ? suggestion.major : "major unavailable"}
```
Comment (MaxLeiter) [inline components/TeammateSuggestions.tsx:45]: Can we show just `submitted` hackers so we won't see the `x unavailable` case?

**Review by MaxLeiter — CHANGES_REQUESTED** [review]

File: components/TeammateSuggestions.tsx (hunk 8 above)
Comment (chloetanlimco) [reply components/TeammateSuggestions.tsx:45]: Yeah, the request filters based on who's confirmed, but I can change it to submitted if that's better

And sorry, yeah there's still a bunch I need to get done, but I'll make those other changes

**Comment by chloetanlimco** [comment]
I'm pretty burnt out at this point, but here's the PR summary: 

Team Management
- Allows anyone on the team to update team visibility
- Allows anyone on the team to update team description
- Displays hackers who have requested to join the team - teams can automatically "accept" these hackers & message
- Displays hackers whom the team has invited to join
- Displays hacker suggestions - teams can message and request hackers


Hacker-Side Not-Part-Of-A-Team
- Allows user to set visibility of profile
- Allows user to set portfolio url
- Displays team invites - hackers can automatically "join" these teams
- Displays teams that the hacker has already requested
- Displays team recommendations - can send requests to the team
- Displays hacker recommendations - LinkedIn <a> shows up if portfolioUrl exists, and hackers are given the option to message (pulls up email)


I could really use some help testing and debugging, if anyone has time. I think setting up the m:n relationship between teams and hackerProfiles messed up the existing team <-> hackerProfile relationship. There are also a couple sequelize routes that are v inelegant, and I probably also missed a few edge cases. 
I'll continue debugging, but idk how much I can get done by Monday. 

image

**Review by abigger87 — APPROVED** [review]

**Pull request merged**